File: static/js/project-management.js
// Project Management JavaScript Functions
document.addEventListener('DOMContentLoaded', function() {
    // Initialize project management features
    initializeProjectForm();
    initializeQuickTaskAdd();
    initializeDashboardRefresh();
});

function initializeProjectForm() {
    const form = document.getElementById('projectForm');
    if (!form) return;
    
    // Set minimum dates
    const today = new Date().toISOString().split('T')[0];
    const startDateInput = document.getElementById('start_date');
    const endDateInput = document.getElementById('end_date');
    
    if (startDateInput) {
        startDateInput.min = today;
        startDateInput.addEventListener('change', function() {
            if (endDateInput) {
                endDateInput.min = this.value;
            }
        });
    }
    
    // Form validation
    form.addEventListener('submit', function(e) {
        const name = document.getElementById('name')?.value;
        const startDate = startDateInput?.value;
        const endDate = endDateInput?.value;
        
        if (!name || !startDate || !endDate) {
            e.preventDefault();
            showAlert('Please fill in all required fields', 'error');
            return;
        }
        
        if (new Date(endDate) <= new Date(startDate)) {
            e.preventDefault();
            showAlert('End date must be after start date', 'error');
            return;
        }
        
        // Show loading state
        const submitBtn = form.querySelector('button[type="submit"]');
        if (submitBtn) {
            submitBtn.disabled = true;
            submitBtn.innerHTML = '<i class="fas fa-spinner fa-spin me-2"></i>Creating...';
        }
    });
}

function initializeQuickTaskAdd() {
    // Quick task addition via modal or inline form
    const quickTaskForms = document.querySelectorAll('.quick-task-form');
    
    quickTaskForms.forEach(form => {
        form.addEventListener('submit', async function(e) {
            e.preventDefault();
            
            const projectId = this.dataset.projectId;
            const taskName = this.querySelector('input[name="task_name"]')?.value;
            const duration = this.querySelector('input[name="duration"]')?.value || 1;
            
            if (!taskName) {
                showAlert('Task name is required', 'error');
                return;
            }
            
            try {
                const response = await fetch('/api/projects/quick-task', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'X-CSRFToken': getCSRFToken()
                    },
                    body: JSON.stringify({
                        project_id: projectId,
                        name: taskName,
                        duration: duration
                    })
                });
                
                const data = await response.json();
                
                if (data.success) {
                    showAlert('Task added successfully!', 'success');
                    this.reset();
                    // Refresh task list if visible
                    refreshProjectTasks(projectId);
                } else {
                    showAlert(data.error || 'Failed to add task', 'error');
                }
                
            } catch (error) {
                console.error('Quick task add error:', error);
                showAlert('Network error. Please try again.', 'error');
            }
        });
    });
}

function initializeDashboardRefresh() {
    const refreshBtn = document.getElementById('refreshDashboard');
    const autoRefreshToggle = document.getElementById('autoRefreshToggle');
    
    if (refreshBtn) {
        refreshBtn.addEventListener('click', function() {
            refreshDashboardData();
        });
    }
    
    if (autoRefreshToggle) {
        autoRefreshToggle.addEventListener('change', function() {
            if (this.checked) {
                startAutoRefresh();
            } else {
                stopAutoRefresh();
            }
        });
        
        // Start auto-refresh if enabled
        if (autoRefreshToggle.checked) {
            startAutoRefresh();
        }
    }
}

async function refreshDashboardData() {
    try {
        const response = await fetch('/api/projects/dashboard-stats');
        const stats = await response.json();
        
        if (stats.error) {
            console.error('Dashboard stats error:', stats.error);
            return;
        }
        
        // Update statistics
        updateStatElement('total-projects', stats.total_projects);
        updateStatElement('active-projects', stats.active_projects);
        updateStatElement('total-tasks', stats.total_tasks);
        updateStatElement('completed-tasks', stats.completed_tasks);
        
        // Update last refresh time
        const now = new Date().toLocaleTimeString();
        const refreshStatus = document.querySelector('.refresh-status');
        if (refreshStatus) {
            refreshStatus.textContent = `Last updated: ${now}`;
        }
        
    } catch (error) {
        console.error('Dashboard refresh error:', error);
    }
}

function updateStatElement(elementId, value) {
    const element = document.getElementById(elementId);
    if (element) {
        element.textContent = value;
        // Add animation
        element.classList.add('stat-updated');
        setTimeout(() => element.classList.remove('stat-updated'), 1000);
    }
}

async function refreshProjectTasks(projectId) {
    try {
        const response = await fetch(`/api/projects/project/${projectId}/tasks`);
        const data = await response.json();
        
        if (data.tasks) {
            // Update task list in the UI
            updateTaskList(data.tasks);
        }
        
    } catch (error) {
        console.error('Task refresh error:', error);
    }
}

function updateTaskList(tasks) {
    const taskList = document.querySelector('.task-list');
    if (!taskList) return;
    
    // Simple task list update - can be enhanced based on UI needs
    const taskItems = tasks.map(task => `
        <div class="task-item" data-task-id="${task.id}">
            <div class="d-flex justify-content-between align-items-center">
                <div>
                    <strong>${task.name}</strong>
                    <small class="text-muted d-block">${task.duration} days</small>
                </div>
                <span class="badge bg-${getStatusBadgeClass(task.status)}">${task.status}</span>
            </div>
        </div>
    `).join('');
    
    taskList.innerHTML = taskItems;
}

function getStatusBadgeClass(status) {
    const statusClasses = {
        'NOT_STARTED': 'secondary',
        'IN_PROGRESS': 'primary',
        'COMPLETED': 'success',
        'ON_HOLD': 'warning',
        'CANCELLED': 'danger'
    };
    return statusClasses[status] || 'secondary';
}

function showAlert(message, type = 'info') {
    // Create alert element
    const alert = document.createElement('div');
    alert.className = `alert alert-${type} alert-dismissible fade show`;
    alert.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
    
    // Add to alerts container or create one
    let alertContainer = document.querySelector('.alerts-container');
    if (!alertContainer) {
        alertContainer = document.createElement('div');
        alertContainer.className = 'alerts-container position-fixed top-0 end-0 p-3';
        alertContainer.style.zIndex = '1055';
        document.body.appendChild(alertContainer);
    }
    
    alertContainer.appendChild(alert);
    
    // Auto-remove after 5 seconds
    setTimeout(() => {
        alert.remove();
    }, 5000);
}

function getCSRFToken() {
    const token = document.querySelector('meta[name=csrf-token]');
    return token ? token.getAttribute('content') : '';
}

// Auto-refresh functionality
let autoRefreshInterval;

function startAutoRefresh() {
    autoRefreshInterval = setInterval(refreshDashboardData, 30000); // 30 seconds
}

function stopAutoRefresh() {
    if (autoRefreshInterval) {
        clearInterval(autoRefreshInterval);
    }
}

File: blueprints/project_management.py
from flask import Blueprint, render_template, request, redirect, url_for, flash, jsonify
from flask_login import login_required, current_user
from datetime import datetime, date
from models import Project, Task, Company, UserRole, TaskStatus, ScheduleType
from extensions import db
import logging

project_mgmt_bp = Blueprint('project_mgmt', __name__)

@project_mgmt_bp.route('/create', methods=['GET', 'POST'])
@login_required
def create_project():
    """Create a new project"""
    if request.method == 'POST':
        try:
            # Get form data
            name = request.form.get('name')
            description = request.form.get('description')
            project_number = request.form.get('project_number')
            start_date = request.form.get('start_date')
            end_date = request.form.get('end_date')
            budget = request.form.get('budget')
            location = request.form.get('location')
            schedule_type = request.form.get('schedule_type', 'GANTT')
            
            # Validation
            if not all([name, start_date, end_date]):
                flash('Name, start date, and end date are required', 'error')
                return render_template('projects/create.html')
            
            # Parse dates
            start_date = datetime.strptime(start_date, '%Y-%m-%d').date()
            end_date = datetime.strptime(end_date, '%Y-%m-%d').date()
            
            if end_date <= start_date:
                flash('End date must be after start date', 'error')
                return render_template('projects/create.html')
            
            # Create project
            project = Project()
            project.name = name
            project.description = description
            project.project_number = project_number
            project.company_id = current_user.company_id
            project.created_by = current_user.id
            project.start_date = start_date
            project.end_date = end_date
            project.budget = float(budget) if budget else None
            project.location = location
            project.schedule_type = ScheduleType[schedule_type]
            
            db.session.add(project)
            db.session.commit()
            
            logging.info(f"New project created: {project.name} by user {current_user.id}")
            flash('Project created successfully!', 'success')
            return redirect(url_for('projects.project_detail', project_id=project.id))
            
        except Exception as e:
            db.session.rollback()
            logging.error(f"Project creation error: {str(e)}")
            flash('Error creating project. Please try again.', 'error')
    
    return render_template('projects/create.html')

@project_mgmt_bp.route('/quick-task', methods=['POST'])
@login_required
def quick_add_task():
    """Quick add task via AJAX"""
    try:
        project_id = request.json.get('project_id')
        task_name = request.json.get('name')
        duration = request.json.get('duration', 1)
        
        if not all([project_id, task_name]):
            return jsonify({'error': 'Project ID and task name required'}), 400
        
        # Verify project access
        project = Project.query.get_or_404(project_id)
        if project.company_id != current_user.company_id:
            return jsonify({'error': 'Access denied'}), 403
        
        # Create task
        task = Task()
        task.name = task_name
        task.project_id = project_id
        task.duration = int(duration)
        task.start_date = date.today()
        task.end_date = date.today()  # Will be calculated properly later
        task.status = TaskStatus.NOT_STARTED
        
        db.session.add(task)
        db.session.commit()
        
        logging.info(f"Quick task added: {task_name} to project {project_id}")
        
        return jsonify({
            'success': True,
            'task': {
                'id': task.id,
                'name': task.name,
                'duration': task.duration,
                'status': task.status.name
            }
        })
        
    except Exception as e:
        db.session.rollback()
        logging.error(f"Quick task creation error: {str(e)}")
        return jsonify({'error': 'Failed to create task'}), 500

@project_mgmt_bp.route('/project/<int:project_id>/tasks')
@login_required
def project_tasks(project_id):
    """Get project tasks as JSON"""
    try:
        project = Project.query.get_or_404(project_id)
        if project.company_id != current_user.company_id:
            return jsonify({'error': 'Access denied'}), 403
        
        tasks = []
        for task in project.tasks:
            tasks.append({
                'id': task.id,
                'name': task.name,
                'start_date': task.start_date.isoformat(),
                'end_date': task.end_date.isoformat(),
                'duration': task.duration,
                'progress': task.progress,
                'status': task.status.name,
                'parent_id': task.parent_task_id
            })
        
        return jsonify({'tasks': tasks})
        
    except Exception as e:
        logging.error(f"Project tasks fetch error: {str(e)}")
        return jsonify({'error': 'Failed to fetch tasks'}), 500

@project_mgmt_bp.route('/dashboard-stats')
@login_required
def dashboard_stats():
    """Get dashboard statistics via API"""
    try:
        stats = {
            'total_projects': Project.query.filter_by(company_id=current_user.company_id).count(),
            'active_projects': Project.query.filter_by(company_id=current_user.company_id, status='active').count(),
            'total_tasks': Task.query.join(Project).filter(Project.company_id == current_user.company_id).count(),
            'completed_tasks': Task.query.join(Project).filter(
                Project.company_id == current_user.company_id,
                Task.status == TaskStatus.COMPLETED
            ).count()
        }
        
        return jsonify(stats)
        
    except Exception as e:
        logging.error(f"Dashboard stats error: {str(e)}")
        return jsonify({'error': 'Failed to fetch stats'}), 500